man/libvdeplug_wirefilter.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "LIBVDEPLUG_WIREFILTER" "1" "April 2023" "VirtualSquare" "Linux User\[aq]s Manual"
.hy
.SH NAME
.PP
\f[C]libvdeplug_wirefilter\f[R] \[en] vdeplug nested module: modify wire
properties
.SH SYNOPSIS
.PP
libvdeplug_wirefilter.so
.SH DESCRIPTION
.PP
Wirefilter is a libvdeplug nested module that allows to manipulate the
properties of the packet flow passing through the plugin.
.PP
This module of libvdeplug4 can be used in any program supporting vde
like \f[C]vde_plug\f[R], \f[C]vdens\f[R], \f[C]kvm\f[R], \f[C]qemu\f[R],
\f[C]user-mode-linux\f[R] and \f[C]virtualbox\f[R].
.PP
The vde_plug_url syntax of this module is the following:
.PP
\ \ \  \f[C]wirefilter://\f[R][ \f[C][\f[R]\f[I]OPTIONS\f[R]\f[C]]\f[R]
]\f[C]{\f[R] \f[I]vde nested url\f[R] \f[C]}\f[R]
.SH OPTIONS
.SS Wire properties
.PP
Wire properties are set with the following format:
.PP
\ \ \ 
[\f[I]direction\f[R]]value[+\f[I]variation\f[R]][\f[I]multiplier\f[R]][\f[I]algorithm\f[R]]
.IP \[bu] 2
\f[I]direction\f[R] specifies the direction of the value (\f[B]LR\f[R]
for left-to-right and \f[B]RL\f[R] for right-to-left, both if not set).
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]variation\f[R] sets a random variation on the value (0 if not set).
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]multiplier\f[R] specifies the unit of measure of the value
(\f[B]K\f[R], \f[B]M\f[R] or \f[B]G\f[R], no multiplier applied by
default).
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]algorithm\f[R] specifies the type of distribution for the variation
(\f[B]U\f[R] for uniform and \f[B]N\f[R] for Gaussian normal, uniform by
default).
.TP
\f[B]\f[CB]delay\f[B]\f[R]
adds extra delay (in milliseconds).
.TP
\f[B]\f[CB]dup\f[B]\f[R]
probability (0-100) of duplicated packets.
Note: 100% causes each packet to be duplicated infinitly.
.TP
\f[B]\f[CB]loss\f[B]\f[R]
probability (0-100) of packets loss.
.TP
\f[B]\f[CB]lostburst\f[B]\f[R]
if not zero, uses Gilbert model for bursty loss.
This is the mean length of lost packet bursts (a two state Markov
chain):
the probability to exit from the faulty state is 1/lostburst and the
probability to
enter the faulty state is loss/(lostburst-(1-loss)).
The loss rate converges to the value of \f[C]loss\f[R].
.TP
\f[B]\f[CB]mtu\f[B]\f[R]
maximum allowed size (in bytes) for packets.
This value is the same for both directions.
.TP
\f[B]\f[CB]bufsize\f[B]\f[R]
maximum size (in bytes) of the packets queue.
Exceeding packets are discarded.
This value is the same for both directions.
.TP
\f[B]\f[CB]bandwidth\f[B]\f[R]
channel bandwidth in bytes/sec.
Sender is not prevented from sending packets, delivery is delayed to
limit the bandwidth to the desired value (like a bottleneck along the
path).
.TP
\f[B]\f[CB]speed\f[B]\f[R]
interface speed in bytes/sec.
Input is blocked for the tramission time of the packet, thus the sender
is prevented from sending too fast.
.TP
\f[B]\f[CB]nofifo\f[B]\f[R]
if set (as flag), it is not guaranteed that packets are delivered in
order (e.g.\ if delayed with different values).
.TP
\f[B]\f[CB]noise\f[B]\f[R]
number of bits damaged/one megabyte.
.SS Blink
.TP
\f[B]\f[CB]blink=path\f[B]\f[R]
if set, a PF_UNIX/DATAGRAM socket is created at the specified path and
for each packet a log message will be sent.
Each log has format:
.RS
.IP
.nf
\f[C]
id direction length (e.g. 6768 LR 44)
\f[R]
.fi
.RE
.TP
\f[B]\f[CB]blinkid=id\f[B]\f[R]
sets the id to be sent for each packet log with \f[C]blink\f[R].
Defaults to Wirefilter pid.
.SS Management
.TP
\f[B]\f[CB]mgmt=path\f[B]\f[R]
creates an unix socket to manage the parameters.
Can be accessed with \f[C]vdeterm\f[R] and used as a remote terminal.
.TP
\f[B]\f[CB]mgmtmode=0700\f[B]\f[R]
access mode of the management socket.
.TP
\f[B]\f[CB]rc=path\f[B]\f[R]
configuration file loaded at Wirefilter startup.
It uses the same syntax of the management interface.
.SS Other
.TP
\f[B]\f[CB]pidfile=path\f[B]\f[R]
saves Wirefilter pid into the specified file.
.SS Markov mode
.PP
Wirefilter provides a more complex set of parameters using a Markov
chain to emulate different states of the link and the transitions
between states.
.PD 0
.P
.PD
Each state is represented by a node.
Markov chain parameters can be set with management commands or rc files
only.
In fact, due to the large number of parameters the command line would
have been unreadable.
.TP
\f[B]\f[CB]markov-numnodes n\f[B]\f[R]
defines the number of different states.
All the parameters of the connection can be defined node by node.
Nodes are numbered starting from zero (to n-1).
.RS
.PP
e.g.:
.IP
.nf
\f[C]
delay 100+10N[4]
\f[R]
.fi
.PP
It is possible to resize the Markov chain at run-time.
New nodes are unreachable and do not have any edge to other states
(i.e.\ each new node has a loopback edge to the node itself with 100%
probability).
When reducing the number of nodes, the weight of the edges towards
deleted nodes is added to the loopback edge.
When the current node of the emulation is deleted, node 0 becomes the
current node.
(The emulation always starts from node 0).
.RE
.TP
\f[B]\f[CB]markov-time ms\f[B]\f[R]
time period (ms) for the markov chain computation.
Each ms microseconds a random number generator decides which is the next
state (default value=100ms).
.TP
\f[B]\f[CB]markov-name n,name\f[B]\f[R]
assign a name to a node of the markov chain.
.TP
\f[B]\f[CB]markov-setnode n\f[B]\f[R]
manually set the current node to the node n.
.TP
\f[B]\f[CB]setedge n1,n2,w\f[B]\f[R]
define an edge between n1 and n2; w is the weight (probability
percentage) of the edge.
The loopback edge (from a node to itself) is always computed as 100%
minus the sum of the weights of outgoing edges.
.TP
\f[B]\f[CB]showedges [ n ]\f[B]\f[R]
list the edges from node n (or from the current node when the command
has no parameters).
Null weight edges are omitted.
.TP
\f[B]\f[CB]showcurrent\f[B]\f[R]
show the current Markov state.
.TP
\f[B]\f[CB]showinfo [ n ]\f[B]\f[R]
show status and information on state (node) n.\ If the parameter is
omitted it shows the status and information on the current state.
.TP
\f[B]\f[CB]markov-debug [ n ]\f[B]\f[R]
set the debug level for the current management connection.
In the actual implementation when n is greater than zero each change of
markov node causes the output of a debug trace.
Debug tracing get disabled when n is zero or the parameter is missing.
.SH EXAMPLES
.PP
Open two terminals.
.PD 0
.P
.PD
In the first terminal run:
.IP
.nf
\f[C]
vdens vxvde://234.0.0.1
ip addr add 10.0.0.1/24 dev vde0
ip link set vde0 up
\f[R]
.fi
.PP
In the second terminal run:
.IP
.nf
\f[C]
vdens wirefilter://[delay=\[dq]LR100\[dq]]{vxvde://234.0.0.1}
ip addr add 10.0.0.2/24 dev vde0
ip link set vde0 up
ping 10.0.0.1
\f[R]
.fi
.PP
Each packet will have a RTT of \[ti]100 ms.
.SH SEE ALSO
.PP
\f[B]vde_plug\f[R](1), \f[B]vdeterm\f[R](1).


libvdeplug_wirefilter.c
/*
 * VDE - libvdeplug_wirefilter
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <limits.h>
#include <libvdeplug.h>
#include <libvdeplug_mod.h>
#include <pthread.h>
#include <poll.h>
#include <time.h>
#include <sys/time.h>
#include <sys/timerfd.h>
#include <wf_conn.h>
#include <wf_queue.h>
#include <wf_time.h>
#include <wf_markov.h>
#include <wf_management.h>
#include <wf_log.h>


#define DROP -1
#define FORWARD 0


static VDECONN *vde_wirefilter_open(char *vde_url, char *descr, int interface_version, struct vde_open_args *open_args);
static ssize_t vde_wirefilter_recv(VDECONN *conn, void *buf, size_t len, int flags);
static ssize_t vde_wirefilter_send(VDECONN *conn, const void *buf, size_t len, int flags);
static int vde_wirefilter_datafd(VDECONN *conn);
static int vde_wirefilter_ctlfd(VDECONN *conn);
static int vde_wirefilter_close(VDECONN *conn);


// Module structure
struct vdeplug_module vdeplug_ops = {
	.vde_open_real = vde_wirefilter_open,
	.vde_recv = vde_wirefilter_recv,
	.vde_send = vde_wirefilter_send,
	.vde_datafd = vde_wirefilter_datafd,
	.vde_ctlfd = vde_wirefilter_ctlfd,
	.vde_close = vde_wirefilter_close
};

static void *packetHandlerThread(void *param);
static void handlePacket(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static void sendPacket(struct vde_wirefilter_conn *vde_conn, Packet *packet);

static char mtuHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static char lossHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static int duplicatesHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static char bufferSizeHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static double bandwidthHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static double speedHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static double delayHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet);
static Packet *noiseHandler(struct vde_wirefilter_conn *vde_conn, Packet *packet);

static int openBlinkSocket(struct vde_wirefilter_conn *vde_conn, char *socket_path);
static int setBlinkId(struct vde_wirefilter_conn *vde_conn, char *id);
static void closeBlink(struct vde_wirefilter_conn *vde_conn);

static int savePidFile(char *path);


static VDECONN *vde_wirefilter_open(char *vde_url, char *descr, int interface_version, struct vde_open_args *open_args) {
	(void)descr; (void)interface_version; (void)open_args;

	init_logs();

	// Random seed
	struct timeval v;
	gettimeofday(&v,NULL);
	srand48(v.tv_sec ^ v.tv_usec ^ getpid());

	struct vde_wirefilter_conn *new_conn = NULL;
	VDECONN *nested_conn;
	char *nested_vnl;
	char *rc_path = NULL;
	char *delay_str = NULL;
	char *dup_str = NULL;
	char *loss_str = NULL;
	char *bursty_loss_str = NULL;
	char *mtu_str = NULL;
	char *nofifo_str = NULL;
	char *channel_size_str = NULL;
	char *bandwidth_str = NULL;
	char *speed_str = NULL;
	char *noise_str = NULL;
	char *blink_path_str = NULL, *blink_id_str = NULL;
	char *management_socket_path = NULL, *management_mode_str = NULL;
	char *pid_file_path = NULL;
	struct vdeparms parms[] = {
		{ "rc", &rc_path },
		{ "delay", &delay_str },
		{ "dup", &dup_str },
		{ "loss", &loss_str },
		{ "lostburst", &bursty_loss_str },
		{ "mtu", &mtu_str },
		{ "nofifo", &nofifo_str },
		{ "bufsize", &channel_size_str },
		{ "bandwidth", &bandwidth_str },
		{ "speed", &speed_str },
		{ "noise", &noise_str },
		{ "blink", &blink_path_str }, { "blinkid", &blink_id_str },
		{ "mgmt", &management_socket_path }, { "mgmtmode", &management_mode_str },
		{ "pidfile", &pid_file_path },
		{ NULL, NULL }
	};

	nested_vnl = vde_parsenestparms(vde_url);											// Gets the nested VNL
	handle_error( vde_parsepathparms(vde_url, parms) != 0, { return NULL; }, NULL );	// Retrieves the plugin parameters
	
	// Opens the connection with the nested VNL
	nested_conn = vde_open(nested_vnl, descr, open_args);
	handle_error( nested_conn == NULL, { return NULL; }, NULL );
	
	new_conn = calloc(1, sizeof(struct vde_wirefilter_conn));
	handle_error( new_conn == NULL, { goto error; }, NULL );
	new_conn->conn = nested_conn;

	// Pipes initialization
	new_conn->send_pipefd = malloc(2*sizeof(int));
	new_conn->receive_pipefd = malloc(2*sizeof(int));
	handle_error( pipe(new_conn->send_pipefd) != 0, { goto error; }, NULL );
	handle_error( pipe(new_conn->receive_pipefd) != 0, { goto error; }, NULL );

	handle_error( initQueue(new_conn, nofifo_str == NULL ? FIFO : NO_FIFO) < 0, { goto error; }, NULL );
	handle_error( initMarkov(new_conn, 1, 0, MS_TO_NS(100)) < 0, { goto error; }, NULL );
	
	setWireValue(new_conn, DELAY, delay_str, 0);
	setWireValue(new_conn, DUP, dup_str, 0);
	setWireValue(new_conn, LOSS, loss_str, 0);
	setWireValue(new_conn, BURSTYLOSS, bursty_loss_str, 0);
	setWireValue(new_conn, MTU, mtu_str, WIRE_BIDIRECTIONAL);
	setWireValue(new_conn, CHANBUFSIZE, channel_size_str, WIRE_BIDIRECTIONAL);
	setWireValue(new_conn, BANDWIDTH, bandwidth_str, 0);
	setWireValue(new_conn, SPEED, speed_str, 0);
	setWireValue(new_conn, NOISE, noise_str, 0);
	new_conn->speed_next[LEFT_TO_RIGHT] = 0;
	new_conn->speed_next[RIGHT_TO_LEFT] = 0;
	new_conn->speed_timer = timerfd_create(CLOCK_REALTIME, 0);
	handle_error( new_conn->speed_timer < 0, { goto error; }, NULL );
	new_conn->bandwidth_next[LEFT_TO_RIGHT] = 0;
	new_conn->bandwidth_next[RIGHT_TO_LEFT] = 0;

	if (blink_path_str) { 
		handle_error( openBlinkSocket(new_conn, blink_path_str) < 0, { goto error; }, NULL );
		handle_error( setBlinkId(new_conn, blink_id_str) < 0, { goto error; }, NULL );
	}

	new_conn->management.socket_fd = -1;
	if (management_socket_path) {
		handle_error( initManagement(new_conn, management_socket_path, management_mode_str) < 0, { goto error; }, NULL );
	}

	if (rc_path) {
		handle_error( loadConfig(new_conn, -1, rc_path) < 0, { goto error; }, NULL );
	}

	if (pid_file_path) {
		handle_error( savePidFile(pid_file_path) < 0, { goto error; }, NULL );
	}

	// Starts packet handler thread
	handle_error( pthread_mutex_init(&new_conn->receive_lock, NULL) != 0, { goto error; }, NULL );
	handle_error( pthread_create(&new_conn->packet_handler_thread, NULL, &packetHandlerThread, (void*)new_conn) != 0, { goto error; }, NULL );

	return (VDECONN *)new_conn;

	error:
		vde_close(nested_conn);
		return NULL;
}

// Right to left
static ssize_t vde_wirefilter_recv(VDECONN *conn, void *buf, size_t len, int flags) {
	/* 
		Note: Packets from the nested plugin (right side) are intercepted by the packet handler thread first
	*/
	(void)len; (void)flags;
	struct vde_wirefilter_conn *vde_conn = (struct vde_wirefilter_conn *)conn;

	const int POLL_SIZE = 1;
	struct pollfd poll_fd[1] = {
		{ .fd=vde_conn->receive_pipefd[0], .events=POLLIN }
	};

	if ( poll(poll_fd, POLL_SIZE, -1) > 0 ) {
		// A packet arrived from the thread (which means that it can be received)
		if (poll_fd[0].revents & POLLIN) {
			ssize_t read_len = read(vde_conn->receive_pipefd[0], buf, VDE_ETHBUFSIZE);
			pthread_mutex_unlock(&vde_conn->receive_lock);
			
			handle_error( read_len < 0, { goto error; }, "Thread error on receive pipe" );

			return read_len;
		}
	}

	error:
		errno = EAGAIN;
		return 1;
}

// Left to right
static ssize_t vde_wirefilter_send(VDECONN *conn, const void *buf, size_t len, int flags) {
	struct vde_wirefilter_conn *vde_conn = (struct vde_wirefilter_conn *)conn;
	uint64_t now = now_ns();

	// Speed delay handling
	if (vde_conn->speed_next[LEFT_TO_RIGHT] > now) {
		usleep( NS_TO_US(vde_conn->speed_next[LEFT_TO_RIGHT] - now) );
	}

	Packet *packet = malloc(sizeof(Packet));
	handle_error( packet == NULL, { goto error; }, "Send packet malloc error" );
	packet->buf = malloc(len);
	handle_error( packet->buf == NULL, { goto error; }, "Send packet payload malloc error" );

	memcpy(packet->buf, buf, len);
	packet->len = len;
	packet->flags = flags;
	packet->direction = LEFT_TO_RIGHT;

	// Passes the packet to the handler
	handle_error( write(vde_conn->send_pipefd[1], (void*)&packet, sizeof(void*)) < 0, { goto error; }, "Thread error on send pipe" );

	return 0;

	error:
		errno = EAGAIN;
		return -1;
}

static int vde_wirefilter_datafd(VDECONN *conn) {
	struct vde_wirefilter_conn *vde_conn = (struct vde_wirefilter_conn *)conn;
	return vde_conn->receive_pipefd[0];
}

static int vde_wirefilter_ctlfd(VDECONN *conn) {
	struct vde_wirefilter_conn *vde_conn = (struct vde_wirefilter_conn *)conn;
	return vde_ctlfd(vde_conn->conn);
}

static int vde_wirefilter_close(VDECONN *conn) {
	struct vde_wirefilter_conn *vde_conn = (struct vde_wirefilter_conn *)conn;

	pthread_cancel(vde_conn->packet_handler_thread);
	pthread_mutex_destroy(&vde_conn->receive_lock);

	close(vde_conn->send_pipefd[0]);
	close(vde_conn->send_pipefd[1]);
	close(vde_conn->receive_pipefd[0]);
	close(vde_conn->receive_pipefd[1]);
	free(vde_conn->send_pipefd);
	free(vde_conn->receive_pipefd);
	close(vde_conn->speed_timer);
	closeQueue(vde_conn);
	closeMarkov(vde_conn);
	if (vde_conn->management.socket_fd > 0) { closeManagement(vde_conn); }
	if (vde_conn->blink.socket_fd > 0) { closeBlink(vde_conn); }
	
	int ret_value = vde_close(vde_conn->conn); // Closes nested connection
	free(vde_conn);
	return ret_value;
}

#define POLL_PIPE_LR 		0
#define POLL_PIPE_RL 		1
#define POLL_QUEUE_TIMER 	2
#define POLL_SPEED_TIMER	3
#define POLL_MARKOV_TIMER	4
#define POLL_MNGM 			5

static void *packetHandlerThread(void *param) {
	pthread_detach(pthread_self());

	struct vde_wirefilter_conn *vde_conn = (struct vde_wirefilter_conn *)param;
	
	const int POLL_SIZE = 6+MNGM_MAX_CONN;
	struct pollfd poll_fd[6+MNGM_MAX_CONN] = {
		{ .fd=vde_conn->send_pipefd[0], .events=POLLIN },					// Left to right packets
		{ .fd=vde_datafd(vde_conn->conn), .events=POLLIN },					// Right to left packets
		{ .fd=vde_conn->queue.timerfd, .events=POLLIN },					// Packet queue timer
		{ .fd=vde_conn->speed_timer, .events=POLLIN },						// Packet speed timer
		{ .fd=vde_conn->markov.timerfd, .events=POLLIN },					// Markov chain state change
		{ .fd=vde_conn->management.socket_fd, .events=POLLIN },				// Management socket
	};
	for (int i=1; i<=MNGM_MAX_CONN; i++) { poll_fd[POLL_MNGM + i].fd = -1; } // Management socket clients

	ssize_t rw_len;
	unsigned char receive_buffer[VDE_ETHBUFSIZE];
	uint64_t now;


	// Starts Markov timer
	setTimer(vde_conn->markov.timerfd, vde_conn->markov.change_frequency);


	while(1) {
		if (poll(poll_fd, POLL_SIZE, -1) > 0) {

			// A packet has to be sent
			if (poll_fd[POLL_PIPE_LR].revents & POLLIN) {
				Packet *packet;
				rw_len = read(vde_conn->send_pipefd[0], &packet, sizeof(void*));
				handle_error( rw_len < 0, {}, "Error while reading send pipe");

				handlePacket(vde_conn, packet);
				packetDestroy(packet);
			}


			// A packet can be received from the nested plugin
			if (poll_fd[POLL_PIPE_RL].revents & POLLIN) {
				now = now_ns();

				// Speed handling
				if (vde_conn->speed_next[RIGHT_TO_LEFT] > now) {
					poll_fd[POLL_PIPE_RL].events &= ~POLLIN; // Stop receiving packets
					setTimer(vde_conn->speed_timer, (vde_conn->speed_next[RIGHT_TO_LEFT] - now));
				}
				else {
					rw_len = vde_recv(vde_conn->conn, receive_buffer, VDE_ETHBUFSIZE, 0);
					handle_error( rw_len < 0, { continue; }, "Error while reading receive pipe");
					
					if (rw_len > 1) { // Not discarded packet
						Packet *packet = malloc(sizeof(Packet));
						handle_error( packet == NULL, { continue; }, "Thread receive packet malloc error");
						packet->buf = malloc(rw_len);
						handle_error( packet->buf == NULL, { continue; }, "Thread receive packet payload malloc error");
						memcpy(packet->buf, receive_buffer, rw_len);
						packet->len = rw_len;
						packet->flags = 0;
						packet->direction = RIGHT_TO_LEFT;

						handlePacket(vde_conn, packet);
						packetDestroy(packet);
					}
				}
			}


			// Time to send something
			if (poll_fd[POLL_QUEUE_TIMER].revents & POLLIN) {
				disarmTimer(vde_conn->queue.timerfd);

				Packet *packet;
				while (vde_conn->queue.size > 0 && nextQueueTime(vde_conn) < now_ns()) {
					packet = dequeue(vde_conn);
					sendPacket(vde_conn, packet);
				}

				// Sets the timer for the next packet
				if (vde_conn->queue.size > 0) {
					setQueueTimer(vde_conn);
				}
			}
			

			// Packets reception (right to left) can be restored (speed handling)
			if (poll_fd[POLL_SPEED_TIMER].revents & POLLIN) {
				disarmTimer(vde_conn->speed_timer);
				poll_fd[POLL_PIPE_RL].events |= POLLIN; // Restart receiving packets
			}


			// Time to change markov chain state
			if (poll_fd[POLL_MARKOV_TIMER].revents & POLLIN) {
				setTimer(vde_conn->markov.timerfd, vde_conn->markov.change_frequency);
				markovStep(vde_conn, vde_conn->markov.current_node);
			}


			// Management socket connection
			if (poll_fd[POLL_MNGM].revents & POLLIN) {
				int new_conn = acceptManagementConnection(vde_conn);

				if (new_conn > 0) {
					poll_fd[POLL_MNGM + vde_conn->management.connections_count].fd = new_conn;
					poll_fd[POLL_MNGM + vde_conn->management.connections_count].events = POLLIN | POLLHUP;
				}
			}

			// Management socket command
			for (int i=1; i<=vde_conn->management.connections_count; i++) {
				if (poll_fd[POLL_MNGM + i].revents & POLLIN) {
					handleManagementCommand(vde_conn, poll_fd[POLL_MNGM + i].fd);
				}
			}

			// Management socket hang-up
			for (int i=1; i<=vde_conn->management.connections_count; i++) {
				if (poll_fd[POLL_MNGM + i].revents & POLLHUP) {
					closeManagementConnection(vde_conn, poll_fd[POLL_MNGM + i].fd);

					// Shifts poll fds
					memmove(&poll_fd[POLL_MNGM+i], &poll_fd[POLL_MNGM+i+1], sizeof(struct pollfd) * (vde_conn->management.connections_count-i));
					poll_fd[POLL_MNGM+vde_conn->management.connections_count].fd = -1;

					vde_conn->management.connections_count--;
					i--;
				}
			}

		}
	}

	pthread_exit(0);
}


static void handlePacket(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	if (mtuHandler(vde_conn, packet) == DROP) { return; }
	if (lossHandler(vde_conn, packet) == DROP) { return; }

	double delay_ms = 0;
	int send_times = 1 + duplicatesHandler(vde_conn, packet);

	for (int i=0; i<send_times; i++) {
		Packet *to_send = packetCopy(packet);
		delay_ms = 0;

		if (bufferSizeHandler(vde_conn, to_send) == DROP) { return; }

		delay_ms += speedHandler(vde_conn, to_send);
		delay_ms += bandwidthHandler(vde_conn, to_send);
		delay_ms += delayHandler(vde_conn, to_send);

		noiseHandler(vde_conn, to_send);

		if (delay_ms > 0 || (vde_conn->queue.fifoness == FIFO && vde_conn->queue.size > 0)) {
			enqueue(vde_conn, to_send, now_ns() + MS_TO_NS(delay_ms));
			setQueueTimer(vde_conn);
		}
		else {
			sendPacket(vde_conn, to_send);
		}
	}
}


/* Sends the packet to the correct destination */
static void sendPacket(struct vde_wirefilter_conn *vde_conn, Packet *packet) {
	ssize_t rw_len;

	// Blink message handling
	if (vde_conn->blink.socket_fd) {
		char *message_content = vde_conn->blink.message + (vde_conn->blink.id_len+1); // Skip id and blank
		
		snprintf(message_content, BLINK_MESSAGE_CONTENT_SIZE, "%s %ld\n",
				(packet->direction == LEFT_TO_RIGHT) ? "LR" : ((packet->direction == RIGHT_TO_LEFT) ? "RL" : "--"), packet->len);		
		sendto(vde_conn->blink.socket_fd, vde_conn->blink.message, strlen(vde_conn->blink.message), 0, 
				(struct sockaddr *)&vde_conn->blink.socket_info, sizeof(vde_conn->blink.socket_info));
	}

	if (packet->direction == LEFT_TO_RIGHT) {
		rw_len = vde_send(vde_conn->conn, packet->buf, packet->len, packet->flags);
		handle_error( rw_len < 0, {}, "Error while sending a LR packet");
	}
	else {
		pthread_mutex_lock(&vde_conn->receive_lock);
		// Makes the packet receivable
		rw_len = write(vde_conn->receive_pipefd[1], packet->buf, packet->len);
		handle_error( rw_len < 0, {}, "Error while sending a RL packet");
	}

	free(packet);
}


static char mtuHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	if (minWireValue(MARKOV_CURRENT(vde_conn), MTU, packet->direction) > 0 && 
		packet->len > minWireValue(MARKOV_CURRENT(vde_conn), MTU, packet->direction)) {
		return DROP;
	}

	return FORWARD;
}

static char lossHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	// Total loss
	if ( minWireValue(MARKOV_CURRENT(vde_conn), LOSS, packet->direction) >= 100.0 ) { return DROP; }

	if (maxWireValue(MARKOV_CURRENT(vde_conn), BURSTYLOSS, packet->direction) > 0) {
		// Loss with Gilbert model
		double loss_val = computeWireValue(MARKOV_CURRENT(vde_conn), LOSS, packet->direction) / 100;
		double burst_len = computeWireValue(MARKOV_CURRENT(vde_conn), BURSTYLOSS, packet->direction);

		switch (vde_conn->bursty_loss_status[packet->direction]) {
			case OK_BURST:
				if ( drand48() < (loss_val / (burst_len*(1-loss_val))) ) { 
					vde_conn->bursty_loss_status[packet->direction] = FAULTY_BURST; 
				}
				break;
			case FAULTY_BURST:
				if ( drand48() < (1.0 / burst_len) ) { 
					vde_conn->bursty_loss_status[packet->direction] = OK_BURST; 
				}
				break;
		}

		if (vde_conn->bursty_loss_status[packet->direction] != OK_BURST) { return DROP; }
	}
	else {
		vde_conn->bursty_loss_status[packet->direction] = OK_BURST;
		
		// Standard loss handling
		if (drand48() < (computeWireValue(MARKOV_CURRENT(vde_conn), LOSS, packet->direction) / 100)) {
			return DROP;
		}
	}

	return FORWARD;
}

static int duplicatesHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	int duplicate_times = 0;

	if (maxWireValue(MARKOV_CURRENT(vde_conn), DUP, packet->direction) > 0) {
		while (drand48() < (computeWireValue(MARKOV_CURRENT(vde_conn), DUP, packet->direction) / 100)) { 
			duplicate_times++; 
		}
	}

	return duplicate_times;
}

static char bufferSizeHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	if (maxWireValue(MARKOV_CURRENT(vde_conn), CHANBUFSIZE, packet->direction)) {
		double buffer_max_size = computeWireValue(MARKOV_CURRENT(vde_conn), CHANBUFSIZE, packet->direction);
		
		if ((vde_conn->queue.byte_size[packet->direction] + packet->len) > buffer_max_size) {
			return DROP;
		}
	}

	return FORWARD;
}

static double bandwidthHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	double delay_ms = 0;

	if (maxWireValue(MARKOV_CURRENT(vde_conn), BANDWIDTH, packet->direction) > 0) {
		double bandwidth = computeWireValue(MARKOV_CURRENT(vde_conn), BANDWIDTH, packet->direction);
		if (bandwidth <= 0) { return DROP; }

		double send_time_ms = (packet->len*1000) / bandwidth;
		uint64_t now = now_ns();

		if (now > vde_conn->bandwidth_next[packet->direction]) {
			// Bandwidth is still below the limit, delay this one to keep the bandwidth up to the limit
			vde_conn->bandwidth_next[packet->direction] = now;
			delay_ms = send_time_ms;
		} else {
			// Bandwidth is overflowing, delay this one until the next bandwidth timestamp 
			double diff = NS_TO_MS( vde_conn->bandwidth_next[packet->direction] - now );
			delay_ms = diff + send_time_ms;
		}
		vde_conn->bandwidth_next[packet->direction] += MS_TO_NS(send_time_ms);
	}

	return delay_ms;
}

static double speedHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	double delay_ms = 0;

	if (maxWireValue(MARKOV_CURRENT(vde_conn), SPEED, packet->direction) > 0) {
		double speed = computeWireValue(MARKOV_CURRENT(vde_conn), SPEED, packet->direction);
		if (speed <= 0) { return DROP; };

		double send_time_ms = (packet->len*1000) / speed;
		uint64_t now = now_ns();

		delay_ms = send_time_ms;
		if (now > vde_conn->speed_next[packet->direction]) {
			vde_conn->speed_next[packet->direction] = now;
		}
		vde_conn->speed_next[packet->direction] += MS_TO_NS(send_time_ms);
	}

	return delay_ms;
}

static double delayHandler(struct vde_wirefilter_conn *vde_conn, const Packet *packet) {
	double delay_ms = 0;

	if (maxWireValue(MARKOV_CURRENT(vde_conn), DELAY, packet->direction) > 0) {
		double delay_value = computeWireValue(MARKOV_CURRENT(vde_conn), DELAY, packet->direction);

		if (delay_value > 0) {
			delay_ms = delay_value;
		}
	}

	return delay_ms;
}

static Packet *noiseHandler(struct vde_wirefilter_conn *vde_conn, Packet *packet) {
	if (maxWireValue(MARKOV_CURRENT(vde_conn), NOISE, packet->direction) > 0) {
		double noise = computeWireValue(MARKOV_CURRENT(vde_conn), NOISE, packet->direction);
		int broken_bits = 0;
		
		// Determines the number of broken bits
		while ((drand48()*8*MEGA) < (packet->len-2)*8*noise) { broken_bits++; }
		
		// Breaks the packet
		for (int i=0; i<broken_bits; i++) {
			int to_flip_bit = drand48() * packet->len*8;
			((char*)packet->buf)[(to_flip_bit >> 3) + 2] ^= 1<<(to_flip_bit & 0x7);
		}
	} 

	return packet;
}


static int openBlinkSocket(struct vde_wirefilter_conn *vde_conn, char *socket_path) {
	vde_conn->blink.socket_info.sun_family = PF_UNIX;
	strncpy(vde_conn->blink.socket_info.sun_path, socket_path, sizeof(vde_conn->blink.socket_info.sun_path)-1);

	vde_conn->blink.socket_fd = socket(PF_UNIX, SOCK_DGRAM, 0);
	handle_error( vde_conn->blink.socket_fd == -1, { return -1; }, "Blink socket error: %s", strerror(errno) );

	return 0;
}

static int setBlinkId(struct vde_wirefilter_conn *vde_conn, char *id) {
	char pid_str[6+1];
	char *to_set_id = id;

	// If id is not set, defaults to pid
	if (to_set_id == NULL) {
		snprintf(pid_str, sizeof(pid_str), "%06d", getpid());
		to_set_id = pid_str;
	}

	vde_conn->blink.id_len = strlen(to_set_id);
	return asprintf(&vde_conn->blink.message, "%s %*c", to_set_id, BLINK_MESSAGE_CONTENT_SIZE, ' ');
}

static void closeBlink(struct vde_wirefilter_conn *vde_conn) {
	close(vde_conn->blink.socket_fd);
	remove(vde_conn->blink.socket_info.sun_path);
}


static int savePidFile(char *path) {
	FILE *fout = NULL;
	
	handle_error( (fout = fopen(path, "w")) == NULL, { return -1; }, "Error in FILE* construction: %s", strerror(errno) );
	handle_error( fprintf(fout, "%d\n", getpid()) < 0, { return -1; }, "Error in writing pidfile" );

	fclose(fout);
	return 0;
}